### PR TITLE
Store the latest cloud provider node addresses

### LINKED ISSUE #65814: Have requests to cloud provider non-blocking
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

There is no guarantee SDK(s) handling requests to cloud providers will response in a reasonable amount of time. Some periodically running routines  (e.g. node status update) can hang for a long time before they finish their job which can cause Kubelet to loose its ability to report/respond in time. E.g. request for node addresses that takes up to 20 minutes forbids the Kubelet to report its Ready state causing the node controller to evict pods from the underlying node.

Instead of relying on individual implementation of SDKs to handle timeouts and request hanging, we should decouple the code responsible for retrieving the node address from a cloud provider and a code responsible for providing the latest node addresses so the node status update routine just reads the recently fetched addresses without any risk of hanging on the abstracted `NodeAddreses` call.

**What you expected to happen**:

The node status update routine call to `NodeAddresses` does not hang for unpredictable amount of time. Otherwise, causing a node to be evicted. Instead, we should implement a go routine running in parallel to kubelet and responsible for fetching the node addresses, storing the recently fetched addresses into a cache. Kubelet will then just ask for the cached node addresses instead of directly talking to cloud provided SDK.

**How to reproduce it (as minimally and precisely as possible)**:

Make sure a response from a cloud provider never returns when asking for node addresses.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

Not filling those as it potentially affects every Kubernetes version on any cloud provider on any OS.

### PR DESCRIPTION
**Pull request by ingvagabund**
**What this PR does / why we need it**:
Buffer the recently retrieved node address so they can be used as soon as the next node status update is run.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #65814

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by ingvagabund** [comment]
@sjenning PTAL

File: pkg/kubelet/kubelet.go
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet.go]: why is this needed here as member var and not just in the kubelet_node_status code block?

File: pkg/kubelet/kubelet_node_status.go
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_node_status.go]: i may be missing something obvious, but seems like cloudProviderLastError can just move next to declaration of var err error above on line 476

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]

**Comment by derekwaynecarr** [comment]
it may help give some clarity on the issue we are having to describe the reason for the change for posterity.

File: pkg/kubelet/kubelet_node_status.go
Comment (ingvagabund) [reply pkg/kubelet/kubelet_node_status.go]: The culprit is actually the `case <-kl.cloudproviderRequestSync` case. Each time the go routine asking for the node addresses finishes, it sends a value to the `kl.cloudproviderRequestSync` channel. If the happens after the `kl.cloudproviderRequestTimeout`, the `kl.cloudproviderRequestSync` is non-empty and in the next node status iteration the select picks the first case `<-kl.cloudproviderRequestSync` instead of waiting for the go routine to finish. Given the `nodeAddresses` and `err` are local variables and are both nil (cause the go routine responsible for setting them stores the values into previous local variables which no longer exist in the `setNodeAddress` scope), the `nodeAddresses` is always nil from the moment the timeout occurs.

This we need to store both (the list of node addresses and the error) into variables that are not local to the `setNodeAddress` method.

Comment (sjenning) [reply pkg/kubelet/kubelet_node_status.go]: there must be a simpler way to do this.  we've added 6 fields to the kubelet struct to support the timeout in this single area of the code.  i'm not saying i see it yet.  but there must be.  i'm too tired to see it atm though.

File: pkg/kubelet/kubelet.go
Comment (ingvagabund) [reply pkg/kubelet/kubelet.go]: Both variables needs to be on the global level so the go routine does not store its values into a local variable that disappears. Plus, the cloud related code is going out at some point and the variables are not exported. So it's fine to keep it that way. We can change it and polish later.

**Comment by sjenning** [comment]
@ingvagabund what would you think about abstracting this to a `cloudRequestManager` that implements
```
type CloudRequestManager interface {
  Start(cloud cloudprovider.Interface) error
  NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.NodeAddress, error)
}
```

and abstract all this internal state out of the kubelet and remove the 6 new fields between https://github.com/kubernetes/kubernetes/pull/62543 and this PR?

`kl.cloudRequestManager.Start()` could be called where we start the other kubelet managers and block starting the kubelet on the success of first `instances.NodeAddresses()`, ensuring that the call in  `setNodeAddress()` never fails. After that, the manager can cache the result for future calls to `kl.cloudRequestManager.NodeAddresses()` and can update against the cloud provided at whatever interval we want (or the cloud provider will allow).

I know this PR has been proven to work, so might do this later.  But I had the idea in my head; wanted to get it in writing for later.

**Comment by ingvagabund** [comment]
@sjenning what you are saying make sense. So far I am aware of two requests that are send:
* request for node addresses

Do we want to have a cloud request manager for each cloud resource we ask for? Or group them based on some criteria? There are cloud resources we ask for once in a while (e.g. the node external ID) and cloud resources we ask for periodically (e.g. the node addresses). So maybe s/`CloudRequestManager`/`CloudRequestSyncManager`/? Effectively build it on top of the informer concept?

File: pkg/kubelet/kubelet_node_status.go
Comment (liggitt) [reply pkg/kubelet/kubelet_node_status.go]: this looks like it is writing to member variables asynchronously outside a lock... isn't that a crashing data race?

**Comment by sjenning** [comment]
> Do we want to have a cloud request manager for each cloud resource we ask for? Or group them based on some criteria? There are cloud resources we ask for once in a while (e.g. the node external ID) and cloud resources we ask for periodically (e.g. the node addresses). So maybe s/CloudRequestManager/CloudRequestSyncManager/? Effectively build it on top of the concept of an informer?

Yes, I think we are on the same wavelength.  CloudRequestSyncManager is good too.  We can use it to basically buffer all cloudprovider requests the kubelet makes since, as we've found out, not all cloudprovider code is equal.

The real question is do you want to go ahead with this PR for now and do that as a follow-on or modify this PR.  I'm on the fence about it.

**Comment by sjenning** [comment]
After some thought, I am in favor of going forward with this fix and doing the refactor in a follow-on PR.
/lgtm

**Comment by liggitt** [comment]
/hold

data race question in https://github.com/kubernetes/kubernetes/pull/65226/files#r198143459 is unresolved

File: pkg/kubelet/kubelet_node_status.go
Comment (ingvagabund) [reply pkg/kubelet/kubelet_node_status.go]: Unlikely to happen but it is. Just the chance is very low. I am addressing this in another PR.

Comment (ingvagabund) [reply pkg/kubelet/kubelet_node_status.go]: Ok, I will update this PR with the suggested refactoring

**Comment by ingvagabund** [comment]
@liggitt @sjenning PTAL

File: pkg/kubelet/cloud_request_manager.go
Comment (sjenning) [inline pkg/kubelet/cloud_request_manager.go]: I don't think we need any of the lastNodeAddress stuff now that we have nodeAddress set async from the from get and it is mutex protected.

Comment (ingvagabund) [reply pkg/kubelet/cloud_request_manager.go]: I removed the original internal go routine completely. It does not make sense to timeout given the request is running in its own go routine as part of the cloud request manager. So basically we are back with the original lines of code. Just wearing a different T-shirt.

File: pkg/kubelet/cloud_request_manager.go
Comment (sjenning) [inline pkg/kubelet/cloud_request_manager.go]: I think we can just remove this and ensure that, in the code, `NodeAddresses()` is never called before `Run()`.  i  don't think we need the `manager.stopCh` at all actually since we are just using `wait.NeverStop`.  Yes, it is more correct to do it this way, but it is done incorrectly in so many places in the kubelet, this is unnecessarily complex.  There is no clean exit path in the kubelet.  We just SIGTERM and everything dies.

**Comment by liggitt** [comment]
fyi, I'm removing one of the early blocking calls to NodeAddresses in https://github.com/kubernetes/kubernetes/pull/65594 in case that helps

**Comment by ingvagabund** [comment]
@liggitt cool, it nicely builds on top of this PR. Once this get merged, you can safely call `NodeAddresses` without any blocking.

**Comment by ingvagabund** [comment]
@dims can you approve?

File: pkg/kubelet/cloud_request_manager.go
Comment (sjenning) [inline pkg/kubelet/cloud_request_manager.go]: the is a persistent condition right? shouldn't this be fatal to the manager?  even the kubelet?

**Comment by dims** [comment]
@ingvagabund the change in `pkg/cloudprovider/providers/fake/fake.go` looks good to me.

/approve

**Comment by ingvagabund** [comment]
@dims thank you very much :)

File: pkg/kubelet/cloud_request_manager.go
Comment (vishh) [inline pkg/kubelet/cloud_request_manager.go]: Is this comment true? The logic below is blocking on cloud provider API call to succeed at least once.

**Review by vishh — COMMENTED** [review]
It looks like this PR is decoupling updating node status from fetching updated node IPs. 
However it is not clear why kubelet needs to keep watching for node addresses continually in the first place.
Also, if there were any production issues that this PR addresses, it would help describe them here in this PR or in a separate issue (preferred).

**Comment by ingvagabund** [comment]
@vishh issue opened https://github.com/kubernetes/kubernetes/issues/65814

File: pkg/kubelet/cloud_request_manager.go
Comment (ingvagabund) [reply pkg/kubelet/cloud_request_manager.go]: The first call to the `NodeAddresses` can be blocking but the remaining ones are not. The assumption is that a node needs to register first before it starts periodically invoking the method. So it is negligible the first call hangs for some time. I can rephrase the statement to be more clear about the fact.

**Comment by ingvagabund** [comment]
@Random-Liu @dchen1107 @yujuhong PTAL

**Comment by sjenning** [comment]
@vishh I agree that it might be unnecessary to poll the cloudprovider for our addresses if we can guarantee that the information we can on the first `NodeAddresses()` call does not change for the life of the kubelet process (or the instance itself).

However, we didn't want to tackle that wider-ranging issue here.  That is very cloudprovider specific and would require acks for all cloud providers saying "yes, our cloudprovider will not return changing return values for `NodeAddresses()` over the life of the kubelet".

This PR is to simply protect the kubelet from latency introduced by the cloudprovider on a NodeAddresses() call.  In our case (Azure), this was a throttling mechanism which could cause the node status update loop to stall and the Node go `NotReady`.

File: pkg/kubelet/kubelet_node_status_test.go
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet_node_status_test.go]: i think we can update this in the future from 10s to 1m, 5m, 10m, etc.

an issue or // todo to track it would be good.

**Review by derekwaynecarr — APPROVED** [review]
/approve

this is an improvement over what exists prior.

i think its up to more debate if the cloud provider interface should be handling this for us by taking a timeout or something on each of its calls.

**Comment by bashofmann** [comment]
Would it make sense to backport this to 1.11 to fix https://github.com/kubernetes/kubernetes/issues/68270 ?

**Comment by alena1108** [comment]
Having correct ip addresses set on the nodes is a plumbing for all subsequent operations on k8s. Would be really helpful to backport this fix to 1.11. We have a lot of users who run 1.11 who are currently facing this problem, and not planning to move to k8s 1.12 yet,

**Comment by openstacker** [comment]
I'm echoing @alena1108 

This is a very critical fix and we have seen this on v1.11.2. Missing internal IP makes the API server can't talk to the worker node and it breaks everything. We also don't want to much v1.12 so fast because I'm afraid there will be new unknown issue.

So please consider to backport this to v1.11.x. Thank you.

**Comment by 2rs2ts** [comment]
We ran into this problem too. Our solution was to just pass the `--node-ip` flag to kubelet.

**Pull request merged**
